lib/trino-parquet/src/main/java/io/trino/parquet/reader/decoders/ValueDecoders.java

```diff
@@ -259,9 +260,12 @@ public static <T> DictionaryDecoder<T> getDictionaryDecoder(
     {
         int size = dictionaryPage.getDictionarySize();
         T dictionary = columnAdapter.createBuffer(size);
-        plainValuesDecoder.init(new SimpleSliceInputStream(dictionaryPage.getSlice()));
+        Slice dictionarySlice = dictionaryPage.getSlice();
+        plainValuesDecoder.init(new SimpleSliceInputStream(dictionarySlice));
         plainValuesDecoder.read(dictionary, 0, size);
-        return new DictionaryDecoder<>(dictionary, columnAdapter);
+        // Since dictionary values use PLAIN encoding, the size in bytes of the dictionary page
+        // is a reasonable estimate of the size of the decoded dictionary retained by DictionaryDecoder
+        return new DictionaryDecoder<>(dictionary, columnAdapter, dictionarySlice.length());
```

Comment:
that seems like an odd way to do it and it doesn't seem to be direct relationship between `dictionary` (of `BufferType`) and `dictionarySlice` (since it has to be decoded).

Can dictionary itself report retained size?

Reply:
The `dictionary` here is a generic buffer like `int[]`, `long[]`, `BinaryBuffer` etc.
We would need to go via ColumnAdapter interface if we want to get size of those structures.
However, dictionary data always uses PLAIN encoding, that's why the raw size of bytes is very close to the size of the buffer that it will be decoded into.

Reply:
> However, dictionary data always uses PLAIN encoding, that's why the raw size of bytes is very close to the size of the buffer that it will be decoded into.

Is it always the case? Can it be significantly understimated (e.g. because of https://stackoverflow.com/questions/24614553/why-is-varint-an-efficient-data-representation) when data contains small numbers?

I see we only have 8 buffer types. Maybe we could make `Buffer` to be interface with retained size method

Reply:
Parquet spec is pretty explicit that dictionary entries are going to be PLAIN encoded https://github.com/apache/parquet-format/blob/master/Encodings.md#dictionary-encoding-plain_dictionary--2-and-rle_dictionary--8
We also assume PLAIN encoding for dictionary in all our existing code for reading dictionaries.
We don't need a `Buffer` interface, I've now added a method on `ColumnAdapter` to explicitly calculate the size of buffer.